transformer_engine/tensorflow/csrc/extensions.cu

```diff
@@ -0,0 +1,1140 @@
+  auto input_cu = MakeNVTETensor(input, input_shape, input_type);
+  auto output_cast_cu = MakeNVTETensor(
+      output_cast, output_cast_shape, output_cast_type, amax, scale, scale_inv);
+  auto output_transpose_cu =
+      MakeNVTETensor(output_transpose, output_transpose_shape,
+                     output_transpose_type, amax, scale, scale_inv);
+
+  nvte_cast_transpose(input_cu.data(), output_cast_cu.data(),
+                      output_transpose_cu.data(), stream);
+}
+
+void dispatch_transpose(void* input,  // i
+                        const std::vector<size_t>& input_shape,
+                        const transformer_engine::DType input_type,
+                        void* output,  // o
+                        const std::vector<size_t>& output_shape,
+                        const transformer_engine::DType output_type,
+                        cudaStream_t stream) {
+  auto input_cu = MakeNVTETensor(input, input_shape, input_type);
+  auto output_cu = MakeNVTETensor(output, output_shape, output_type);
+
+  nvte_transpose(input_cu.data(), output_cu.data(), stream);
+}
+
+void dispatch_bgrad_cast_transpose_fusion(
+    void* input,  // i
+    const std::vector<size_t>& input_shape,
+    const transformer_engine::DType input_type,
+    void* scale,  // i
+    const std::vector<size_t>& scale_shape,
+    const transformer_engine::DType scale_type,
+    void* cast_output,  // o
+    const std::vector<size_t>& cast_output_shape,
+    const transformer_engine::DType cast_output_type,
+    void* transposed_output,  // o
+    const std::vector<size_t>& transposed_output_shape,
+    const transformer_engine::DType transposed_output_type,
+    void* amax,  // o
+    const std::vector<size_t>& amax_shape,
+    const transformer_engine::DType amax_type,
+    void* dbias,  // o
+    const std::vector<size_t>& dbias_shape,
+    const transformer_engine::DType dbias_type,
+    void* scale_inv,  // o
+    const std::vector<size_t>& scale_inv_shape,
+    const transformer_engine::DType scale_inv_type, cudaStream_t stream) {
+  auto input_cu = MakeNVTETensor(input, input_shape, input_type);
+  auto cast_output_cu = MakeNVTETensor(
+      cast_output, cast_output_shape, cast_output_type, amax, scale, scale_inv);
+  auto transposed_output_cu =
+      MakeNVTETensor(transposed_output, transposed_output_shape,
+                     transposed_output_type, amax, scale, scale_inv);
+  auto dbias_cu = MakeNVTETensor(dbias, dbias_shape, dbias_type);
+  transformer_engine::TensorWrapper workspace;
+
+  nvte_cast_transpose_dbias(input_cu.data(), cast_output_cu.data(),
+                            transposed_output_cu.data(), dbias_cu.data(),
+                            workspace.data(), stream);
+
+  // Fill workspace
+  auto w_s = workspace.shape();
+  std::vector<size_t> w_shape_vec{w_s.data, w_s.data + w_s.ndim};
+
+  void* workspace_ptr = AllocateSpace(w_shape_vec, workspace.dtype());
```

Comment:
Does the TensorFlow GPU allocator do buffer caching like in PyTorch? I'm a little suspicious since the stream logic only affects how we call `cudaMemset`. If we're allocating a new GPU buffer every time we call this operation, that would add a lot of overhead. It's not critical for the first pass since it won't affect correctness, but we'll need to figure this out when we want performance.

There are a few other dispatch functions in this file where this is relevant.

Reply:
TF GPU allocator allocates a large chunk of GPU memory and suballocates it using a best-fit with coalescing (BFC) allocator, which is quite efficient on "creating" and "freeing" memory. We have already checked the profiling results and didn't notice significant overhead on this.